internal/search/search.go

```diff
@@ -0,0 +1,220 @@
+				hunksForFlag = append(hunksForFlag, *match)
+			}
+		}
+	}
+	return hunksForFlag
+}
+
+func (f file) toHunks(projKey string, aliases map[string][]string, ctxLines int, delimiters string) *ld.ReferenceHunksRep {
+	hunks := []ld.HunkRep{}
+	for flagKey, flagAliases := range aliases {
+		hunks = append(hunks, f.aggregateHunksForFlag(projKey, flagKey, flagAliases, ctxLines, delimiters)...)
+	}
+	if len(hunks) == 0 {
+		return nil
+	}
+	return &ld.ReferenceHunksRep{Path: f.path, Hunks: hunks}
+}
+
+// mergeHunks combines the lines and aliases of two hunks together for a given file
+// if the hunks do not overlap, returns each hunk separately
+// assumes the startingLineNumber of a is less than b and there is some overlap between the two
+func mergeHunks(a, b ld.HunkRep) []ld.HunkRep {
+	if a.StartingLineNumber > b.StartingLineNumber {
+		a, b = b, a
+	}
+
+	aLines := strings.Split(a.Lines, "\n")
+	bLines := strings.Split(b.Lines, "\n")
+	overlap := a.Overlap(b)
+	// no overlap
+	if overlap < 0 || len(a.Lines) == 0 && len(b.Lines) == 0 {
+		return []ld.HunkRep{a, b}
+	} else if overlap >= len(bLines) {
+		// subset hunk
+		return []ld.HunkRep{a}
+	}
+
+	combinedLines := append(aLines, bLines[overlap:]...)
+	return []ld.HunkRep{
+		{
+			StartingLineNumber: a.StartingLineNumber,
+			Lines:              strings.Join(combinedLines, "\n"),
+			ProjKey:            a.ProjKey,
+			FlagKey:            a.FlagKey,
+			Aliases:            helpers.Dedupe(append(a.Aliases, b.Aliases...)),
+		},
+	}
+}
+
+// processFiles starts goroutines to process files individually. When all files have completed processing, the references channel is closed to signal completion.
+func processFiles(files chan file, references chan ld.ReferenceHunksRep, projKey string, aliases map[string][]string, ctxLines int, delimiters string) {
+	w := new(sync.WaitGroup)
+	for file := range files {
+		file := file
+		w.Add(1)
+		go func() {
+			reference := file.toHunks(projKey, aliases, ctxLines, delimiters)
+			if reference != nil {
+				references <- *reference
+			}
+			w.Done()
+		}()
+	}
+	w.Wait()
+	close(references)
+}
+
+func SearchForRefs(projKey, workspace string, aliases map[string][]string, ctxLines int, delimiters string) ([]ld.ReferenceHunksRep, error) {
+	files := make(chan file)
+	references := make(chan ld.ReferenceHunksRep)
+
+	// Start workers to process files asynchronously as they are written to the files channel
+	go processFiles(files, references, projKey, aliases, ctxLines, delimiters)
+
+	// Blocks until all files have been read, but not necessarily processed
+	err := readFiles(files, workspace)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := []ld.ReferenceHunksRep{}
+
+	defer sort.SliceStable(ret, func(i, j int) bool {
+		return ret[i].Path < ret[j].Path
+	})
+
+	totalHunks := 0
+	for reference := range references {
+		ret = append(ret, reference)
+
+		// Reached maximum number of files with code references
+		if len(ret) >= maxFileCount {
+			return ret, nil
```

Comment:
if this returns early, what happens to the background goroutine writing into this channel?